File: 1036-rotting-oranges/rotting-oranges.py
from queue import Queue

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = Queue()
        r = len(grid)
        c = len(grid[0])

        vis = set()

        for i in range(r):
            for j in range(c):
                if grid[i][j] == 2:
                    q.put((i,j,0))
                    vis.add((i,j))
        
        drow = [1,-1,0,0]
        dcol = [0,0,1,-1]

        ans = 0
        cnt = 0
        while not q.empty():
            i,j,ct = q.get()
            cnt += 1
            for k in range(4):
                ci = i + drow[k]
                cj = j + dcol[k]

                if (ci >= 0 and ci < r and cj >= 0 and cj < c) and ((ci,cj) not in vis) and (grid[ci][cj] == 1):
                    #print("inside if")
                    q.put((ci,cj,ct+1))
                    vis.add((ci,cj))
                    ans = ct+1
        
        
        
        
        
        for i in range(r):
            for j in range(c):
                if grid[i][j] == 1 and ((i,j) not in vis):
                    return -1
        
        return ans



        

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<pair<int, int>, int>> q;
        
        int r = grid.size();
        int c = grid[0].size();

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (grid[i][j] == 2) {
                    q.push({{i, j}, 0});
                }
            }
        }

        int lastTime = 0;
        
        vector<int> drow = {1, -1, 0, 0};
        vector<int> dcol = {0, 0, 1, -1};

        while (!q.empty()) {
            int i = q.front().first.first;
            int j = q.front().first.second;
            int time = q.front().second;
            q.pop();

            for (int k = 0; k < 4; k++) {
                int ci = i + drow[k];
                int cj = j + dcol[k];

                if (ci >= 0 && ci < r && cj >= 0 && cj < c && grid[ci][cj] == 1) {
                    grid[ci][cj] = 2;
                    q.push({{ci, cj}, time + 1});
                }
            }

            lastTime = max(time, lastTime);
        }

        // Check if there's any fresh orange left
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (grid[i][j] == 1) {
                    return -1; // If any fresh orange is still left
                }
            }
        }

        return lastTime;
    }
};
